src/Microsoft.DotNet.Build.Tasks/ParseTestCoverageInfo.cs

```diff
@@ -0,0 +1,219 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Xml;
+using System.Collections.Generic;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public class ParseTestCoverageInfo : Task
+    {
+        // Path to the directory that contais *.coverage.xml coverage info.
+        [Required]
+        public string CoverageInfoDirectory { get; set; }
+
+        // One per xml parsed, key = test name, value list of modules visited
+        private Dictionary<string, Dictionary<string, ModuleInfo>> testCoverageInfo = null;
+
+        public override bool Execute()
+        {
+            if (Directory.Exists(CoverageInfoDirectory))
+            {
+                Log.LogMessage(MessageImportance.Normal, "Parsing coverage files...");
+                if (!ProcessDirectory(CoverageInfoDirectory))
+                {
+                    Log.LogError("There was no coverage file found.");
+                    return false;
+                }
+            }
+            else
+            {
+                Log.LogError("{0} is not a valid path", CoverageInfoDirectory);
```

Comment:
One suggestion would be to move your error cases to checks at the beginning. The same would be true for the missing coverage file case. You'd have less nested if's and the success path would be succinct.